### ISSUE TITLE
Swagger response type other than string in swagger-node (express) fails validation

### ISSUE THREAD

**Issue by dgamache** [swagger-api/swagger-node#378]
I cannot seem to handle response types other than string.  If I try to return a boolean, I get this:
```
  project started here: http://localhost:10010/
  project will restart on changes.
  to restart at any time, enter `rs`
try this:
curl http://127.0.0.1:10010/hello?name=Scott
Error: Response validation failed: not a valid boolean: "true"
    at throwErrorWithCode (C:\dev\swagger\hello-world\node_modules\swagger-express-mw\node_modules\swagger-node-runner\node_modules\swagger-tools\lib\validators.js:121:13)
    at validateTypeAndFormat (C:\dev\swagger\hello-world\node_modules\swagger-express-mw\node_modules\swagger-node-runner\node_modules\swagger-tools\lib\validators.js:538:7)
    at Object.module.exports.validateSchemaConstraints (C:\dev\swagger\hello-world\node_modules\swagger-express-mw\node_modules\swagger-node-runner\node_modules\swagger-tools\lib\validators.js:630:7)
    at validateValue (C:\dev\swagger\hello-world\node_modules\swagger-express-mw\node_modules\swagger-node-runner\node_modules\swagger-tools\middleware\swagger-validator.js:117:16)
    at ServerResponse.res.end (C:\dev\swagger\hello-world\node_modules\swagger-express-mw\node_modules\swagger-node-runner\node_modules\swagger-tools\middleware\swagger-validator.js:252:9)
    at ServerResponse.send (C:\dev\swagger\hello-world\node_modules\express\lib\response.js:204:10)
    at ServerResponse.json (C:\dev\swagger\hello-world\node_modules\express\lib\response.js:249:15)
    at isDoorOpen (C:\dev\swagger\hello-world\api\controllers\hello_world.js:51:7)
    at swaggerRouter (C:\dev\swagger\hello-world\node_modules\swagger-express-mw\node_modules\swagger-node-runner\node_modules\swagger-tools\middleware\swagger-router.js:407:20)
    at swagger_router (C:\dev\swagger\hello-world\node_modules\swagger-express-mw\node_modules\swagger-node-runner\fittings\swagger_router.js:31:5)
```
My real-life API is far more complex and wants to return booleans, numbers in different formats and bmp and jpeg image types.  Le's say for simplicity that I want an /isDoorOpen endpoint to return a boolean.

I duplicate steps here: [swagger-api/swagger-node](https://github.com/swagger-api/swagger-node).
I use express middleware
I added this YAML:
```YAML
   /isDoorOpen:
    # binds a127 app logic to a route
    x-swagger-router-controller: hello_world
    get:
      description: Returns 'true' to the caller because my door is always open
      # used as the method name of the controller
      operationId: isDoorOpen
      responses:
        "200":
          description: Success
          schema:
            type: boolean
        # responses may fall through to errors
        default:
          description: Error
          schema:
            $ref: "#/definitions/ErrorResponse"
```
And in the hello_world.js file I added these two sections:
```JavaScript
module.exports = {
  hello: hello,
  isDoorOpen: isDoorOpen
};
```
and
```JavaScript
function isDoorOpen(req, res) {
  var isOpen = true; // Details omitted
  var doorStatus = util.format('%s', isOpen);

  res.json(doorStatus);
}
```
What do you suggest that I do.  The validator certain should work for this case, right?  Thank you.

**Comment by garywong89** [comment]
I edited the yaml file to include a $ref after the boolean response.
'
$ref: "#/definitions/HelloWorldResponse"
'

https://gist.github.com/garywong89/99c6f670901bb653641fd09510c2d7d6

**Comment by garywong89** [comment]
Hi dgamache, I believe that some the formatting of the variable doorStatus is causing part of the error by formatting it into a string first before returning back a boolean. I included a link of a gist. See if that helps resolve the issue.

https://gist.github.com/garywong89/99c6f670901bb653641fd09510c2d7d6

**Comment by dgamache** [comment]
That did help that issue garywong89.  Thank you.  I need to review further to make sure I understand why, but it seems like the utility function changed the formatting from what is desired.  I went on to update my application and ran into odd behavior around getting images.  Please allow me to update the sample, and even if I solve the issue, I want to post it for the greater good of the community.

**Comment by dgamache** [comment]
Hi garywong89.  Removing the formatting made the boolean response work correctly.  To illustrate what worked corectly for string, boolean and image responses, I included a gist:

[https://gist.github.com/dgamache/12cb46ce5b47e22a2915d9420e34627d.js](https://gist.github.com/dgamache/12cb46ce5b47e22a2915d9420e34627d.js)

This completes the original intent of my post, but leaves me with a new question.
When I defined the API, I found it useful to define some "non-200" responses.  For example, if I request a picture and none is there, I would want to return a 404 response code.  I have been experimenting, reading reference and seeking examples and I cannot find how to return that response from the middleware successfully.  Part of the problem is that the type is a file, mime type image/bmp but when an error happens in the middleware and I want to return a 404, that is not a file.  Instead it's an object or a string at least.  It seems likely that I would have to change the response type dynamically at run-time.  I have not figured out how to do that though. 

What do you think?  (also should we close this issue and open up this as a new issue?)

Thank you,
David

**Comment by garywong89** [comment]
Hi dgamache, so I was tesing the code ran into the same error you were facing. As you have pointed out the file was location a string and returning the string location regardless if the file was there. 

I believe the issue for the validation should be coming from the controller's response. I added a gist on the bottom that should resolve the issue and allow you to change response status code if needed. I added a module File System as a helper to check if the file is present first before sending the response back. Please tell me if this helps resolve some of the issue you are having?

https://gist.github.com/garywong89/47d30822d31a2b145fd472f39343c771

**Comment by dgamache** [comment]
Hi @garywong89 , I did try a variety of methods to change the response status code and have not found one that works correctly yet.

First, I tried your gist but found the response did not end:
`res.status(404);`
So I chained an end to it:
`res.status(404).end();`
This resulted in:

> {"message":"Response validation

output in the browser.  I also tried a few other methods but kept coming up with the Response validation error.

The YAML does not do much more that list the response code as one that could occur:
```
"404":
    description: "Picture is missing."
```

How do I get this to respond with a 404 status?  It seems like this would be a rather common, useful thing to do, I just haven't figured out how yet.

**Comment by dgamache** [comment]
Does anyone have any feedback on how to return different responses like 404 from the middleware?

**Comment by garywong89** [comment]
Hi @dgamache, not too sure but I tried passing back the argument in the hello_world.js file. It seems like the default the response for images is 200, so in order to display the error I think the server needs to get passed a response to know what output to expect.
https://gist.github.com/garywong89/09cab56ae5ad87c429875abee4c42eb2

**Comment by oliverhr** [comment]
@dgamache you have to use `res.statusCode`
```
res.statusCode = 404;
return next(err);
```

**Comment by dgamache** [comment]
Hi @garywong89 and @oliverhr, I tried both approaches.  Evidently, the server needed to be passed a response composed like this in express:

> 
        // sends the response back with the status you wish the response to be located in
        var myMessage = util.format('File not found went to error')
        res.json(myMessage).status("404");

I tried your recommendation too @oliverhr, but I wasn't able to get it to work as intended.

In all, I think this issue (mostly an issue with my understanding best practices) is resolved.

Thank you,
David
